ref(js): Rewrite outdated "Transaction Name" documentation (#12091)
getsentry/sentry-docs: docs/platforms/javascript/common/enriching-events/transaction-name/index.mdx

---
title: Transaction Name

> description: "Learn how to set or override the transaction name to improve issue and span grouping."

Author: ```suggestion
description: "Learn how to set or override the transaction name to improve issue and trace grouping."
```

supported:
  - javascript
notSupported:
  - javascript.cordova
  - javascript.nextjs
---

The Sentry SDK sets a "transaction name" which is used to annotate error events with their point of failure.

> Furthermore, if you use <PlatformLink to="/tracing/">Tracing</PlatformLink>, Sentry refers to the root spans of your application as a "transaction" and groups traces based on the root span's "transaction" name.

Author: I went with slightly different wording but your general structure, thanks!

Sentry.getCurrentScope().setTransactionName("UserListView");
```

The transaction name on the scope only applies to error events, not the active root span. Similarly, if an error happens during an active span, the span's name won’t be applied to the error's transaction name.

### Setting the Name on the Event

You can use `beforeSend` to set the transaction name on the event:

```javascript {3}
Sentry.init({
  beforeSend(event) {
    event.transaction = "UserListView";
    return event;
  },
});
```

## Setting the Root Span Name

There are multiple options for overriding the root span name (this name shows up in the Sentry UI, for example when searching for traces or when in performance insights):

<PlatformCategorySection supported={['browser']}>

### At Root Span Start

Sentry's `browserTracingIntegration` automatically sets the transaction name based on the current route or URL for all `pageload` and `navigation` transactions.
To override the transaction name, use the <PlatformLink to="/tracing/instrumentation/automatic-instrumentation/#beforestartspan">`beforeStartSpan` callback</PlatformLink> of the integration.

</PlatformCategorySection>

### While the Root Span Is Active

> _Available starting with: v8.44.0_

Author: ```suggestion
_Available starting with: v8.46.0_
```

> _Available starting with: v8.46.0_

Author: ```suggestion
_Available starting with: v8.47.0_
```

Ideally, the transaction name does not contain variable values such as user
IDs but has rather low cardinality while still uniquely identifying a piece of
code you care about.
For example, instead of `'GET /api/users/123/details'`, set `'GET /api/users/:id/details'`.
This ensures that errors or traces for all requests made to this route are grouped together.

### When to set the Transaction Name

> 
> We generally recommend to let the SDK set the transaction name automatically.
> However, in some cases, the SDK might not be able to set a name or the set name doesn't work for your use case.
> 
> If you need to set the transaction name manually, we recommend setting it _as early as possible_ in your application code, specifically when the operation you want to describe with the transaction name starts.
> This ensures that errors thrown early in the life cycle of your application are correctly annotated.
> Furthermore, it ensures consistency across a trace if you use [distributed tracing](/concepts/key-terms/tracing/distributed-tracing/).
> 
> For example, in a server application, you might set the transaction name in a middleware that runs before the request is processed.
> In a web application, you might set the transaction name when you enter a new route (e.g. in an SPA Router).
> 

Reviewer: ```suggestion

It's usually best to let the SDK set the transaction name automatically. However, if it can't or the name doesn't suit your needs, you can set it manually.

We recommend setting it _as early as possible_ in your application code, ideally when the operation you want to track starts. This ensures that errors thrown early in the application lifecycle are annotated correctly and keeps traces consistent if you're using [distributed tracing](/concepts/key-terms/tracing/distributed-tracing/).

For example, in a server app, you can set the transaction name in middleware before handling a request.
In a web app, you can set it when navigating to a new route, like in an SPA router.

```

## Setting the Error Transaction Name

> There are several ways to set the transaction name for error events sent to Sentry.

Reviewer 2: In "Setting the Root Span Name", you mention where this name is displayed and used for. You could include this information here as well.

### Setting the Name on the Scope

Set the transaction name on the current scope when the operation you want to group starts:

```javascript
Sentry.getCurrentScope().setTransactionName("UserListView");
```

> Note that the transaction name on the scope is only applied to error events. It does not influence the name of a potentially active root span.
> Likewise, if a span is active while an error is thrown, its name will not be applied to the error event's transaction name.

Reviewer: ```suggestion
The transaction name on the scope only applies to error events, not the active root span. Similarly, if an error happens during an active span, the span's name won’t be applied to the error's transaction name.
```

### Setting the Name on the Event

You can use `beforeSend` to set the transaction name on the event:

```javascript {3}
Sentry.init({
  beforeSend(event) {
    event.transaction = "UserListView";
    return event;
  },
});
```

## Setting the Root Span Name

> To override the name of the root span name (i.e. the trace name or transaction name in the Sentry UI), you have multiple options:

Reviewer 2: Maybe specify where in the Sentry UI it is used.

> To override the name of the root span name (i.e. the trace name or transaction name in the Sentry UI), you have multiple options:
> 

Reviewer: ```suggestion
There are multiple options for overriding the root span name (for example, the trace or transaction name in the Sentry UI):

```

<PlatformCategorySection supported={['browser']}>

### At Root Span Start

Sentry's `browserTracingIntegration` automatically sets the transaction name based on the current route or URL for all `pageload` and `navigation` transactions.

> If you want to override the transaction name, you can do so in the <PlatformLink to="/tracing/instrumentation/automatic-instrumentation/#beforestartspan">`beforeStartSpan` callback</PlatformLink> of the integration.
> 

Reviewer: ```suggestion
To override the transaction name, use the <PlatformLink to="/tracing/instrumentation/automatic-instrumentation/#beforestartspan">`beforeStartSpan` callback</PlatformLink> of the integration.

```

</PlatformCategorySection>

> ### While the root span is active

Reviewer: ```suggestion
### While the Root Span Is Active
```

> _Available since: v8.44.0_

Reviewer: ```suggestion
_Available starting with: v8.44.0_
```

> Sometimes, for example if you have a router that only emits the route change after it occurred, you might not be able to set the final root span name at span start but a little bit later. In this case, you can update the root span name while the span is active:

Reviewer: ```suggestion
Sometimes, (for example if you have a router that only emits the route change after it occurred) you might not be able to set the final root span name at span start. In this case, you can update the root span name while the span is active:
```

```javascript
const activeSpan = Sentry.getActiveSpan();
const rootSpan = activeSpan && Sentry.getRootSpan(activeSpan);

Sentry.updateSpanName(rootSpan, "UserListView");
```

Learn more about <PlatformLink to="/tracing/instrumentation/custom-instrumentation/#updating-the-span-name">updating the span name</PlatformLink>.

> ### After the root span finished

Reviewer: ```suggestion
### After the Root Span Has Finished
```

> If you cannot determine the root span name before the span finishes, you can retroactively change it in `beforeSendTransaction`:

Reviewer: ```suggestion
If you can't determine the root span name before the span finishes, you'll be able to change it retroactively in `beforeSendTransaction`:
```

```javascript
Sentry.init({
  beforeSendTransaction(event) {
    event.transaction = "UserListView";
    return event;
  },
});
```

> We recommend to only do this, if you cannot use any of the other options to set the root span name earlier.

Reviewer: ```suggestion
<Note>
Only do this if you can't set the root span name earlier using other options.
</Note>
```

## Further Information

> You might be wondering why these two types of transaction names exist and why they are set independently.
> The reason is mostly historic evolvement of the Sentry product and the SDKs.
> The error transaction name existed long before Sentry provided tracing capabilities and where it was only applied to error events to better group issues.

Reviewer: ```suggestion

You might wonder why there are two separate transaction names and why they're set independently. This is mainly due to Sentry's history and evolution. The error transaction name existed before Sentry offered tracing, and it was used solely to group error events. When tracing was introduced, the root span in a span tree was also called a "transaction," with its own name.
```

> In the first iteration of Sentry's tracing or performance monitoring product, we decided to call the root span of a span tree within an application a "Transaction", which also had a name.
> In the second, now ongoing, tracing iteration, Sentry's new UI features no longer mention "Transactions" for root spans but rather just "spans" and "traces".
> However, the two concepts (error transaction name and root span name) are still ambiguous and used so throughout various older parts of the Sentry UI.

Reviewer: ```suggestion
In later iterations, Sentry shifted focus in the UI from "transactions" to "spans" and "traces," but the older concepts remain in parts of the UI. The SDKs adapted their APIs, moving away from transaction-based designs. For example, the `setTransactionName` API on the `Scope` refers to error transaction names, not spans. Similarly, a span name doesn’t automatically apply to error events, keeping the two concepts [separate](https://github.com/getsentry/sentry-javascript/issues/10846).
```

> With this product evolvement in mind, the SDKs also had to adapt their APIs around tracing, specifically moving away from transaction-based APIs in version 8.
> Consequently, the SDK exposes the `setTransactionName` API on the `Scope` which—although the name might suggest differently from historic context—has nothing to do with spans.
> Likewise, the span name is not automatically applied to error events either, meaning both concepts are completely [decoupled](https://github.com/getsentry/sentry-javascript/issues/10846) from each other.

Reviewer: ```suggestion

```

> We believe that the ambiguity of these APIs will decrease over time as the Sentry product further moves away from associating the "Transaction" term with tracing and spans.

Reviewer: ```suggestion
As Sentry continues to evolve, the ambiguity between these terms is expected to decrease as the product moves away from associating "transactions" with tracing and spans.
```
I understand why you added it, but I'm not sure this section is necessary. I think it might add more confusion. But let me know if you disagree :)
